### PR TITLE
Adding data source support for google_compute_instance_group_manager

### PR DESCRIPTION
**Pull request by googlerjk**
Adding support for data source to google_compute_instance_group_manager resource
[fixes this git issue]( https://github.com/hashicorp/terraform-provider-google/issues/7222)

If this PR is for Terraform, I acknowledge that I have:
- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**
```release-note:new-datasource
`google_compute_instance_group_manager`
```

### CODE REVIEW

File: mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+Get a Compute Instance Group Manager within GCE.
+For more information, see [the official documentation](https://cloud.google.com/compute/docs/instance-groups#managed_instance_groups)
+and [API](https://cloud.google.com/compute/docs/reference/latest/instanceGroupManagers)
+
+## Example Usage
+
+```hcl
+data "google_compute_instance_group_manager" "igm1" {
+  name = "k8s1-abcdef01-myns-mysvc-8080-4b6bac43"
```
Comment (c2thorn) [inline mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown:18]: Why is this value used? It isn't wrong in any sense, just confusing a bit? why not "my-igm" or something simple

File: mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceGoogleComputeInstanceGroupManager() *schema.Resource {
+
+	return &schema.Resource{
+		Read: dataSourceComputeInstanceGroupManagerRead,
+		Schema: map[string]*schema.Schema{
```
Comment (c2thorn) [inline mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go:14]: Why not use the `datasourceSchemaFromResourceSchema` method here? How did you choose these fields to add?

**Review by c2thorn — CHANGES_REQUESTED** [review]

**Review by c2thorn — CHANGES_REQUESTED** [review]
Also, the release note type is misspelled and the release note is not formatted correctly. empty lines break the bot I believe

File: mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go (hunk 2 above)
Comment (googlerjk) [reply mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go]: using resource schema now.

File: mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown (hunk 1 above)
Comment (googlerjk) [reply mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown]: updated

File: mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
+package google
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestGoogleComputeInstanceGroupManager(t *testing.T) {
```
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go:9]: This should be prefixed `TestAccDataSource`.

File: mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceGoogleComputeInstanceGroupManager() *schema.Resource {
+
```
Comment (c2thorn) [inline mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go:11]: You likely still need to add the fields `name`, `project`, `self_link`, and `zone`

**Review by c2thorn — CHANGES_REQUESTED** [review]
The test failed. Did you test `TestGoogleComputeInstanceGroupManager` ?

File: mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go (hunk 4 above)
Comment (googlerjk) [reply mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go:11]: Added the fields

File: mmv1/products/compute/api.yaml [hunk 5]
```diff
@@ -7528,7 +7530,7 @@ objects:
         description: |
           TargetPool resources to which instances in the instanceGroup field are
           added. The target pools automatically apply to all of the instances in
-          the managed instance group.
+          the managed instance group.'
```
Comment (c2thorn) [inline mmv1/products/compute/api.yaml:7533]: Why is this added?

File: mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		"templateName": "tf-tst-templt" + randString(t, 6),
+		"autoHealName": "tf-tst-ah-name" + randString(t, 6),
+	}
+
+	vcrTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceGoogleComputeInstanceGroupManager_basic(context),
```
Comment (c2thorn) [inline mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go:29]: Can we add another test step to confirm that using the zone/name works too?

File: mmv1/products/compute/api.yaml (hunk 5 above)
Comment (c2thorn) [reply mmv1/products/compute/api.yaml:7533]: Also, technically no changes to this api.yaml are needed at all. we could consider removing all of the changes to this file just for simplicity's sake.

**Review by c2thorn — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go (hunk 6 above)
Comment (googlerjk) [reply mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go:29]: ok

File: mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+		name = "%{igmName}"
+		base_instance_name = "%{baseName}"
+		zone               = "us-central1-a"
+
+		version {
+		  instance_template  = google_compute_instance_template.igm-basic.id
+		  name = "primary"
+		}
+
+		all_instances_config {
```
Comment (shuyama1) [inline mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go:71]: `all_instances_config` is beta-only feature. Consider remove this from the generate test file or add add version guard around the code that should only be in beta like
```
<% unless version == 'ga' -%>
// test case + config here
<% end -%>
```
Otherwise, test will fail in GA

**Review by shuyama1 — DISMISSED** [review]
Tested it locally and it doesn't look like we throw any errors if the resource is not found

File: mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go (hunk 7 above)
Comment (googlerjk) [reply mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go]: this property has been removed.

File: mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+		service_account {
+			scopes = ["userinfo-email", "compute-ro", "storage-ro"]
+		}
+	}`, context)
+}
+
+func testAccDataSourceGoogleComputeInstanceGroupManager_basic2(context map[string]interface{}) string {
+	return Nprintf(`
+	resource "google_compute_health_check" "autohealing" {
```
[the remark this answers is not in the document]
Comment (googlerjk) [reply mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go]: all tabs have been removed.

File: mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+		if err := d.Set("name", parsed.Name); err != nil {
+			return fmt.Errorf("Error setting name: %s", err)
+		}
+		if err := d.Set("zone", parsed.Zone); err != nil {
+			return fmt.Errorf("Error setting zone: %s", err)
+		}
+		if err := d.Set("project", parsed.Project); err != nil {
+			return fmt.Errorf("Error setting project: %s", err)
+		}
+		d.SetId(fmt.Sprintf("projects/%s/zones/%s/instanceGroupManagers/%s", parsed.Project, parsed.Zone, parsed.Name))
```
[the remark this answers is not in the document]
Comment (googlerjk) [reply mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go:37]: @c2thorn is this going to be a problem?

File: mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+* `fingerprint` - The fingerprint of the instance group manager.
+
+* `instance_group` - The full URL of the instance group created by the manager.
+
+* `description` - Textual description of the instance group manager.
+
+* `self_link` - The URI of the resource.
+
+* `status` - The status of this managed instance group.
```
Comment (c2thorn) [inline mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown:55]: Instead of writing out all of the fields here, we can just refer to the resource version of the documentation for all exported attributes. This will save time later, since if the resource gets more fields added, we won't have to update the data source documentation as well. 

An example of other data sources doing this is at https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/d/cloudfunctions2_function.html.markdown#attributes-reference

**Review by c2thorn — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown (hunk 10 above)
Comment (googlerjk) [reply mmv1/third_party/terraform/website/docs/d/compute_instance_group_manager.html.markdown]: resolved

File: mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+		}
+
+		network_interface {
+			network = "default"
+		}
+
+		service_account {
+			scopes = ["userinfo-email", "compute-ro", "storage-ro"]
+		}
+	}`, context)
```
[the remark this answers is not in the document]
Comment (googlerjk) [reply mmv1/third_party/terraform/tests/data_source_google_compute_instance_group_manager_test.go]: resolved

**Review by c2thorn — APPROVED** [review]

File: mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+		project, err := getProject(d, config)
+		if err != nil {
+			return err
+		}
+		d.SetId(fmt.Sprintf("projects/%s/zones/%s/instanceGroupManagers/%s", project, zone, name.(string)))
+	} else {
+		return errors.New("Must provide either `self_link` or `zone/name`")
+	}
+
+	return resourceComputeInstanceGroupManagerRead(d, meta)
```
Comment (c2thorn) [inline mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go:52]: @shuyama1 found in the case where a user sets an invalid IGM name the datasource silently fails instead of erroring out. This is because the `resourceComputeInstanceGroupManagerRead` method does not return an error when the API 404's but simply sets the ID to "". This is fine in the resource case, but the data source should error out.

To handle this, we can check to see if the ID is "" _after_ calling resourceComputeInstanceGroupManagerRead.
That means we'll want to save the result of resourceComputeInstanceGroupManagerRead as a variable, return the variable if it is not nil, otherwise check the ID and return an error if it is "".

**Review by c2thorn — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go (hunk 12 above)
Comment (googlerjk) [reply mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go:52]: :+1:  Working on this now

Comment (googlerjk) [reply mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go]: I imagine we are expecting to be empty or working. If its not "" do I return the Errors object of the non nil object returned?

Comment (c2thorn) [reply mmv1/third_party/terraform/data_sources/data_source_google_compute_instance_group_manager.go]: something like
```
err := resourceComputeInstanceGroupManagerRead(d, meta)

if err != nil {
	return err
}
if d.Id() == "" {
	return errors.New("Instance Manager Group not found")
}
return nil
```

**Review by c2thorn — APPROVED** [review]
